src/controllers/vstatefulset_controller/trusted/rely_guarantee.rs

```diff
@@ -0,0 +1,246 @@
+    |s: ClusterState| {
+        &&& req.preconditions is Some
+        &&& req.preconditions->0.resource_version is Some
+        &&& {
+            let etcd_obj = s.resources()[req.key];
+            let owner_references = etcd_obj.metadata.owner_references->0;
+            &&& s.resources().contains_key(req.key)
+            &&& etcd_obj.metadata.resource_version is Some
+            &&& etcd_obj.metadata.resource_version
+                == req.preconditions->0.resource_version
+            &&& etcd_obj.metadata.owner_references is Some
+            &&& exists |vsts: VStatefulSetView| #[trigger] owner_references.contains(vsts.controller_owner_ref())
+        }
+    }
+}
+
+// Other controllers don't try to delete a pod matching a VSTS
+pub open spec fn interfere_delete_pvc_req(req: DeleteRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        &&& req.preconditions is Some
+        &&& req.preconditions->0.resource_version is Some
+        &&& {
+            let etcd_obj = s.resources()[req.key];
+            let owner_references = etcd_obj.metadata.owner_references->0;
+            &&& s.resources().contains_key(req.key)
+            &&& etcd_obj.metadata.resource_version is Some
+            &&& etcd_obj.metadata.resource_version
+                == req.preconditions->0.resource_version
+            &&& etcd_obj.metadata.name is Some
+            &&& exists |vsts: VStatefulSetView| pvc_name_match(etcd_obj.metadata.name->0, vsts)
+        }
+    }
+}
+
+pub open spec fn interfere_get_then_delete_req(req: GetThenDeleteRequest) -> bool {
+    &&& req.key.kind == PodView::kind() ==> interfere_get_then_delete_pod_req(req)
+    // GetThenDelete on PVC will fail because PVC owned by VSTS in etcd has
+}
+
+// Other controllers don't try to delete pod owned by a VSTS.
+pub open spec fn interfere_get_then_delete_pod_req(req: GetThenDeleteRequest) -> bool {
+    &&& req.owner_ref.controller is Some
+    &&& req.owner_ref.controller->0
+    &&& req.owner_ref.kind != VStatefulSetView::kind()
+}
+
+// VSTS Guarantee Condition
+
+pub open spec fn vsts_guarantee(controller_id: int) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        forall |msg| {
+            &&& #[trigger] s.in_flight().contains(msg)
+            &&& msg.content is APIRequest
+            &&& msg.src.is_controller_id(controller_id)
+        } ==> match msg.content->APIRequest_0 {
+            APIRequest::CreateRequest(req) => vsts_guarantee_create_req(req),
+            APIRequest::GetThenUpdateRequest(req) => vsts_guarantee_get_then_update_req(req),
+            APIRequest::GetThenDeleteRequest(req) => vsts_guarantee_get_then_delete_req(req),
+            // Update, UpdateStatus and Delete requests do not guarantee
+            _ => true,
+        }
+    }
+}
+
+// VSTS controller only creates Pods owned by itself
+// and only creates PVC matching its PVC templates
+pub open spec fn vsts_guarantee_create_req(req: CreateRequest) -> bool {
```

Comment:
I think you reused the guarantee conditions of VRS? They don't apply to VSTS. Also, the guarantee condition should say that VSTS's create request always has the right name prefix, similar to the rely condition